webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/batch/CleanPoliciesFolder.scala

```diff
@@ -0,0 +1,141 @@
+* GNU General Public License for more details.
+*
+* You should have received a copy of the GNU General Public License
+* along with Rudder.  If not, see <http://www.gnu.org/licenses/>.
+
+*
+*************************************************************************************
+*/
+
+package com.normation.rudder.batch
+
+import java.nio.file.FileSystemException
+
+import better.files.File
+import better.files.File.root
+import com.normation.inventory.domain.NodeId
+import com.normation.rudder.domain.logger.ScheduledJobLogger
+import com.normation.rudder.domain.nodes.NodeInfo
+import com.normation.rudder.services.nodes.NodeInfoService
+import monix.execution.Scheduler.{global => scheduler}
+import net.liftweb.common._
+
+import scala.concurrent.duration._
+import scala.util.{Success, Failure => Catch}
+import scala.util.Try
+
+
+/**
+ * A naive scheduler which checks every updateInterval if software needs to be deleted
+ */
+class CleanPoliciesFolder(
+     nodeInfoService : NodeInfoService
+   , updateInterval  : FiniteDuration
+) {
+
+  val logger = ScheduledJobLogger
+
+
+  if (updateInterval < 1.hour) {
+    logger.info(s"Disable automatic unreferenced policies directory (update interval cannot be less than 1 hour)")
+  } else {
+    logger.debug(s"***** starting batch that purge unreferenced policies directory, every ${updateInterval.toString()} *****")
+    scheduler.scheduleWithFixedDelay(1.hour, updateInterval) {
+
+
+    }
+  }
+
+  def launch = {
+    (for {
+      nodes <- nodeInfoService.getAll()
+      d <- Box.tryo{cleanPoliciesFolderOfDeletedNode(nodes)}
+      deleted <- d
+    } yield {
+      deleted
+    }) match {
+      case eb: EmptyBox =>
+        val error = (eb ?~! s"Error when deleting unreferenced policies directory")
+        logger.error(error.messageChain)
+        error.rootExceptionCause.foreach(ex =>
+          logger.debug("Exception was:", ex)
+        )
+        error
+      case Full(deleted) =>
+        logger.info(s"Deleted ${deleted.length} unreferenced policies folder")
+        if (logger.isDebugEnabled && deleted.length > 0)
+          logger.debug(s"Deleted policies folder of Nodes: ${deleted.mkString(",")}")
+        Full(deleted)
+    }
+  }
+
+  def cleanPoliciesFolderOfDeletedNode(currentNodes : Map[NodeId,NodeInfo]) : Box[List[String]] = {
+    import File._
+
+    def getNodeFolders(file: File, parentId: String): Iterator[String] = {
+      file.children.flatMap {
+        nodeFolder =>
+          val nodeId = NodeId(nodeFolder.name)
+          if (
+            currentNodes.get(nodeId) match {
+              case None => true
+              case Some(nodeInfo) =>
+                val policyServerId = nodeInfo.policyServerId
+                parentId != policyServerId.value
+            }
+          ) {
+            nodeFolder.delete()
+            val res = Iterator(nodeFolder.name)
+            res
+          } else {
+            if ((nodeFolder / "share").exists) {
```

Comment:
Can you add a comment here to explain why you can't just rely on the "isPolicyServer" info? Something along "if the node was a relay and is not anymore, we still want to delete the old unused "share" folder"

Reply:
maybe it has no share folder too, (no nodes below, no share folder generated, it makes an error if a relay tries to update and there is no share)

Maybe i can do something like if (node.isPolicyServer && nodeFolder / "share") to not check existence of this folder if the node is not a policyServer